core/src/main/scala/org/bykn/bosatsu/rankn/Infer.scala

```diff
@@ -500,20 +502,20 @@ object Infer {
                   (exists.map(_._1).iterator.zip(ms) ++
                     univ.map(_._1).iterator.zip(sksT.iterator)).toMap
                 )
-                (sks2, ms2, ty) <- loop(ty1, path)
+                (sks2, ms2, ty) <- loop(ty1, allCo)
               } yield (sks1 ::: sks2, ms ::: ms2, ty)
             } else pure((Nil, Nil, t))
 
           case ta @ Type.TyApply(left, right) =>
             // Rule PRFUN
             // we know the kind of left is k -> x, and right has kind k
             // since left: Rho, we know loop(left, path)._3 is Rho
-            (varianceOfCons(ta, region), loop(left, path))
+            (varianceOfCons(ta, region), loop(left, allCo))
               .flatMapN { case (consVar, (sksl, el, ltpe0)) =>
                 // due to loop invariant
                 val ltpe: Type.Rho = ltpe0.asInstanceOf[Type.Rho]
-                val rightPath = consVar * path
-                loop(right, rightPath)
+                val allCoRight = allCo && (consVar == Variance.co)
```

Comment:
I don't *really* know why this works, but it is skolemizing in a smaller set of cases from before, yet still keeping functions in weak-prenex form (which isn't type checked, but is asserted in the inference and typechecking process).

So, I guess one argument is we want to skolemize the least we can to make inference work, especially with function return types, and this is sufficient to do it. It also passes all the current tests.

The paper I based the type inference on doesn't have a type system as complex as bosatsu, so I had to generalize it, and in this case I suppose my guess was wrong (which was refusing to typecheck things that seem like they should, such as the examples in this PR).